### PR TITLE
Auto-register function via `method_added` hook

### PR DESCRIPTION
**Pull request by splattael**
Less setup ceremony (`module_function` and `register_from`).

:warning: This PR probably kills kittens. A LOT

/cc @solnic @AMHOL

### CODE REVIEW

File: lib/transproc.rb [hunk 1]
```diff
@@ -23,6 +17,13 @@ def [](name)
   def functions
     @_functions ||= {}
   end
+
+  module AutoRegister
```
Comment (solnic) [inline lib/transproc.rb]: `FunctionContainer`?
`FunctionProvider`?
`FunctionDefinitions`?

hmmmm????

Comment (splattael) [inline lib/transproc.rb]: Just "Functions"?

Comment (solnic) [inline lib/transproc.rb]: +1!

**Comment by AMHOL** [comment]
Sorry, I still prefer the `register_from` "no magic" way myself, plus this way will define twice as many methods, each will be defined as both an instance method and a class method.

**Comment by splattael** [comment]
@AMHOL I think the amount of methods definitions won't change.

**Comment by splattael** [comment]
@AMHOL 

## RE: amount of methods

Adding
```ruby
    p :module => methods(false).size
    p :instance => private_instance_methods(false).size
```
at the end of `Transproc::ArrayTransformations` module shows for both `master` and this PR the same result:
```ruby
{:module=>4}
{:instance=>4}
```

## RE: `register_from`

We can keep `Transproc.register_from` to inject transproc functions from outside. Currently, I do not see a use-case except the removed `Transproc.register_from(self)`.

This might be a use-case:
```ruby
module MyFuncs
  def proc1(value)
  end

  def proc2(value)
  end
end

Transproc.register_from(MyFuncs)
```

For the same functionality I would prefer
```ruby
module MyFuncs
  extend Transproc::Functions

  def proc1(value)
  end

  def proc2(value)
  end
end
```
(but that's just me)

**Comment by AMHOL** [comment]
Yeah, to be fair the `extend` method does look a little nicer, I'm just not keen on `method_defined` hook, but can live with it I guess lol

**Comment by AMHOL** [comment]
:+1: looks good

**Comment by solnic** [comment]
I think this is a good change because it feels more natural in Ruby to do it like that. Otherwise we might be violating [POLS](http://en.wikipedia.org/wiki/Principle_of_least_astonishment), I think.

Merging in.

**Pull request merged**